### PR TITLE
[SYCLomatic] Refine the logic to remove wait

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/Utility.cpp [hunk 1]
```diff
@@ -3464,62 +3472,124 @@ bool analyzeMemcpyOrder(
...
+            Result.getNodeAs<DeclRefExpr>("VarReference");
+        if (!MatchedDRE) {
+          continue;
+        }
+        auto D = dyn_cast_or_null<clang::VarDecl>(MatchedDRE->getDecl());
+        if (!D) {
+          continue;
+        }
+        bool IsGlobalVar = GlobalVDSet.count(D);
+        if (dpct::DpctGlobalInfo::getTypeName(D->getType()) == "cudaError_t") {
```
Comment (ziranzha) [inline clang/lib/DPCT/Utility.cpp]: Check the type at first.

File: clang/lib/DPCT/Utility.cpp [hunk 2]
```diff
@@ -3726,6 +3792,43 @@ bool canOmitMemcpyWait(const clang::CallExpr *CE) {
     DREOffsetVec = Iter->second.DREOffsetVec;
   }
 
+  auto CheckIfDstImpactNextCE = [&](const CallExpr *CurCE,
```
Comment (ziranzha) [inline clang/lib/DPCT/Utility.cpp]: Can we check it in above logic?

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/Utility.cpp [hunk 3]
```diff
@@ -3435,87 +3461,158 @@ bool analyzeMemcpyOrder(
...
+                        clang::dpct::DpctGlobalInfo::getContext());
+      }
+      for (auto &Result : Results) {
+        const DeclRefExpr *MatchedDRE =
+            Result.getNodeAs<DeclRefExpr>("VarReference");
+        if (!MatchedDRE) {
+          continue;
+        }
+        auto D = dyn_cast_or_null<clang::VarDecl>(MatchedDRE->getDecl());
+        if (!D) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.cpp]: this function is quite big now,  ~300 line.  Difficult to understand the framework of this function, please refactor the code.

File: clang/test/dpct/USM-restricted.cu [hunk 4]
```diff
@@ -470,25 +470,25 @@ int foo2() {
...
-  // CHECK: q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size).wait();
+  // CHECK: q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size);
   cudaMemcpyToSymbol(constData, h_A, size, 1);
-  // CHECK: q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size).wait();
+  // CHECK: q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size);
   cudaMemcpyToSymbol("constData", h_A, size, 1);
   // CHECK: errorCode = DPCT_CHECK_ERROR(q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size));
   errorCode = cudaMemcpyToSymbol(constData, h_A, size, 1);
-  // CHECK: MY_SAFE_CALL(DPCT_CHECK_ERROR(q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size).wait()));
+  // CHECK: MY_SAFE_CALL(DPCT_CHECK_ERROR(q_ct1.memcpy((char *)(constData.get_ptr()) + 1, h_A, size)));
```
Comment (zhimingwang36) [inline clang/test/dpct/USM-restricted.cu:479]: what is the migration result if h_A is updated b/w two API.

File: clang/lib/DPCT/Utility.cpp [hunk 5]
```diff
@@ -3330,25 +3340,149 @@ const clang::Stmt *getBodyofAncestorFCStmt(const clang::Stmt *S) {
   return CS;
 }
 
+/// Find all related DREs with the declarations in \p VDSet. Except the DREs in
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.cpp]: Please update the comment by more detail or relationship b/w args.

**Review by zhimingwang36 — CHANGES_REQUESTED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**